// vue3-form-wizard@0.2.9/dist/vue3-form-wizard.es.js
/*
 * vue3-form-wizard
 * Creator:parsa jiravand
 * vue3-form-wizard is a vue based component with no external depenendcies which simplifies tab wizard management.
 * v0.2.9
 * MIT License
 */

import { defineComponent as H, createElementBlock as u, openBlock as d, renderSlot as v, computed as i, normalizeClass as T, createElementVNode as g, normalizeStyle as k, unref as s, toDisplayString as z, ref as p, provide as ee, watch as te, onMounted as le, withKeys as K, createCommentVNode as O, Fragment as ge, renderList as ke, createVNode as W, normalizeProps as w, guardReactiveProps as $, withCtx as U, createTextVNode as G, inject as ae, getCurrentInstance as me, onBeforeUnmount as Ce, withDirectives as Te, vShow as pe } from "vue";
const ze = {
  class: "wizard-btn",
  tabindex: "-1",
  type: "button"
}, q = /* @__PURE__ */ H({
  setup(t) {
    return (o, r) => (d(), u("button", ze, [
      v(o.$slots, "default")
    ]));
  }
}), Ie = ["tabindex", "id", "aria-controls", "aria-disabled", "aria-selected"], Be = ["innerHTML"], we = ["innerHTML"], se = /* @__PURE__ */ H({
  props: {
    tab: null,
    transition: { default: "" },
    index: { default: 0 },
    disableBackOnClickStep: { type: Boolean, default: !1 }
  },
  setup(t) {
    const o = t, r = i(() => ({
      backgroundColor: o.tab.color
    })), m = i(() => o.tab.active ? {} : { color: o.tab.color }), a = i(() => ({
      borderColor: o.tab.color
    })), b = i(() => ({
      borderColor: o.tab.errorColor,
      backgroundColor: o.tab.errorColor
    })), C = i(() => ({
      color: o.tab.validationError ? o.tab.errorColor : o.tab.color
    })), l = i(() => o.tab.shape === "square"), h = i(() => o.tab.shape === "tab"), S = i(
      () => o.disableBackOnClickStep ? "cursor: default" : ""
    );
    return (y, L) => (d(), u("li", {
      class: T({ active: t.tab.active })
    }, [
      g("a", {
        class: T({ disabled: !t.tab.checked }),
        style: k(s(S))
      }, [
        g("div", {
          class: T(["wizard-icon-circle md", {
            checked: t.tab.checked,
            square_shape: s(l),
            tab_shape: s(h)
          }]),
          role: "tab",
          tabindex: t.tab.checked ? 0 : "",
          id: `step-${t.tab.tabId}`,
          "aria-controls": t.tab.tabId,
          "aria-disabled": t.tab.active,
          "aria-selected": t.tab.active,
          style: k([
            t.tab.checked ? s(a) : {},
            t.tab.validationError ? s(b) : {}
          ])
        }, [
          t.tab.active ? (d(), u("div", {
            key: 0,
            class: T(["wizard-icon-container", { square_shape: s(l), tab_shape: s(h) }]),
            style: k([s(r), t.tab.validationError ? s(b) : {}])
          }, [
            v(y.$slots, "active-step", {}, () => [
              t.tab.customIcon ? (d(), u("span", {
                key: 0,
                class: "wizard-icon",
                innerHTML: t.tab.customIcon
              }, null, 8, Be)) : (d(), u("i", {
                key: 1,
                class: T([t.tab.icon ? t.tab.icon : "", "wizard-icon"]),
                style: k(t.tab.checked ? s(m) : "")
              }, z(t.tab.icon ? null : t.index + 1), 7))
            ])
          ], 6)) : v(y.$slots, "default", { key: 1 }, () => [
            t.tab.customIcon ? (d(), u("span", {
              key: 0,
              class: "wizard-icon",
              innerHTML: t.tab.customIcon
            }, null, 8, we)) : (d(), u("i", {
              key: 1,
              class: T([t.tab.icon ? t.tab.icon : "", "wizard-icon"]),
              style: k(t.tab.checked ? s(m) : "")
            }, z(t.tab.icon ? null : t.index + 1), 7))
          ])
        ], 14, Ie),
        v(y.$slots, "title", {}, () => [
          g("span", {
            class: T(["stepTitle", { active: t.tab.active, has_error: t.tab.validationError }]),
            style: k([t.tab.active || t.tab.checked ? s(C) : {}, { "margin-top": "5px" }])
          }, z(t.tab.title), 7)
        ]),
        v(y.$slots, "customIcon")
      ], 6)
    ], 2));
  }
});
function $e() {
  return document.activeElement?.id || "";
}
function ne(t = []) {
  const o = $e();
  return t.findIndex((m) => m.tabId === o);
}
function oe(t) {
  const o = document.getElementById(t);
  o && o.focus();
}
function Se(t) {
  return t && typeof t.then == "function";
}
const xe = ["id", "onKeyup"], Ee = {
  key: 0,
  class: "wizard-header"
}, Ke = { class: "wizard-title" }, Oe = { class: "category" }, Pe = { class: "wizard-navigation" }, Ae = {
  key: 0,
  class: "wizard-progress-with-circle"
}, Le = { class: "wizard-tab-content" }, Me = {
  key: 1,
  class: "wizard-card-footer clearfix"
}, Re = {
  key: 0,
  class: "wizard-footer-left"
}, Fe = ["onKeyup"], Ne = { class: "wizard-footer-right" }, We = ["onKeyup"], qe = ["onKeyup"], He = /* @__PURE__ */ H({
  props: {
    id: { default: void 0 },
    title: { default: "Awesome Wizard" },
    subtitle: { default: "Split a complicated flow in multiple steps" },
    nextButtonText: { default: "Next" },
    backButtonText: { default: "Back" },
    finishButtonText: { default: "Finish" },
    hideButtons: { type: Boolean, default: !1 },
    validateOnBack: { type: Boolean, default: !1 },
    color: { default: "#e74c3c" },
    errorColor: { default: "#8b0000" },
    shape: { default: "circle" },
    layout: { default: "horizontal" },
    stepsClasses: { default: "" },
    stepSize: { default: "md" },
    transition: { default: "" },
    startIndex: { default: 0 },
    disableBackOnClickStep: { type: Boolean, default: !1 },
    disableBack: { type: Boolean, default: !1 }
  },
  emits: {
    "on-change": (t, o) => !0,
    "update:startIndex": (t) => !0,
    "on-complete": () => !0,
    "on-loading": (t) => !0,
    "on-error": (t) => !0,
    "on-validate": (t, o) => !0
  },
  setup(t, { emit: o }) {
    const r = t, m = i(() => r.id || "fw_" + (/* @__PURE__ */ new Date()).valueOf()), a = p(0), b = p(0), C = p(!1), l = p([]), h = i(() => l.value.length), S = i(() => a.value === h.value - 1), y = i(() => r.layout === "vertical"), L = i(() => a.value !== 0), P = i(() => 1 / (h.value * 2) * 100), I = i(() => {
      let e = 0;
      return a.value > 0 ? e = P.value * (a.value * 2 + 1) : e = P.value, e;
    }), J = i(() => ({
      backgroundColor: r.color,
      width: `${I.value}%`,
      color: r.color
    })), M = i(() => ({
      backgroundColor: r.color,
      borderColor: r.color,
      color: "white"
    })), B = i(() => ({
      nextTab: A,
      prevTab: V,
      activeTabIndex: a.value,
      isLastStep: S.value,
      fillButtonStyle: M.value
    })), Q = (e, n) => {
      o("on-change", e, n), o("update:startIndex", n);
    }, ie = (e) => {
      const n = h.value;
      e.tabId = `${e.title.replace(/ /g, "")}${n}`, l.value.splice(n, 0, e), n < a.value + 1 && (b.value = n, x(a.value + 1, n));
    }, ce = (e) => {
      const n = l.value.indexOf(e);
      n > -1 && (n === a.value && (b.value = a.value - 1, x(a.value, a.value - 1)), n < a.value && (b.value = a.value - 1, a.value = a.value - 1, Q(a.value + 1, a.value)), l.value.splice(n, 1));
    }, R = (e) => {
      const n = e > a.value;
      if (e <= b.value) {
        const c = () => {
          n && e - a.value > 1 ? (x(a.value, a.value + 1), N(a.value, c)) : (x(a.value, e), Y(a.value));
        };
        n ? N(a.value, c) : (F(null), c());
      }
      return e <= b.value;
    }, A = () => {
      const e = () => {
        a.value < h.value - 1 ? (x(a.value, a.value + 1), Y(a.value)) : o("on-complete");
      };
      N(a.value, e);
    }, V = () => {
      const e = () => {
        a.value > 0 && (F(null), x(a.value, a.value - 1));
      };
      r.validateOnBack ? N(a.value, e) : e();
    }, re = () => {
      const e = ne(l.value);
      if (e !== -1 && e < l.value.length - 1) {
        const n = l.value[e + 1];
        n.checked && oe(n.tabId);
      }
    }, ue = () => {
      const e = ne(l.value);
      if (e !== -1 && e > 0) {
        const n = l.value[e - 1].tabId;
        oe(n);
      }
    }, D = (e) => {
      C.value = e, o("on-loading", e);
    }, F = (e) => {
      l.value[a.value].validationError = e, o("on-error", e);
    }, de = (e, n) => {
      F(null), Se(e) ? (D(!0), e.then((c) => {
        D(!1), X(c === !0, n);
      }).catch((c) => {
        D(!1), F(c);
      })) : X(e === !0, n);
    }, X = (e, n) => {
      o("on-validate", e, a.value), e ? n() : l.value[a.value].validationError = "error";
    }, N = (e, n) => {
      if (C.value)
        return;
      const c = l.value[e];
      if (c && c.beforeChange !== void 0) {
        const f = c.beforeChange();
        de(f, n);
      } else
        n();
    }, Y = (e) => {
      if (C.value)
        return;
      const n = l.value[e];
      n && n.afterChange !== void 0 && n.afterChange();
    }, x = (e, n, c = !0) => {
      const f = l.value[e], E = l.value[n];
      return f && (f.active = !1), E && (E.active = !0), c && a.value !== n && Q(e, n), a.value = n, j(a.value), !0;
    }, ve = (e) => {
      console.warn("Route navigation needs to be handled by parent component in Composition API");
    }, be = (e) => {
      let n = -1;
      const c = l.value.find((f, E) => {
        const _ = f.route === e;
        return _ && (n = E), _;
      });
      c && !c.active && (n > a.value, R(n));
    }, fe = () => {
      l.value.forEach((e) => {
        e.active = !1;
      });
    }, Z = (e) => {
      fe();
      const n = l.value[e];
      n && (n.active = !0, n.checked = !0, ve());
    }, j = (e) => {
      Z(e), e > b.value && (b.value = e), a.value = e;
    }, he = () => {
      l.value.length > 0 && r.startIndex === 0 && Z(a.value), r.startIndex < l.value.length ? j(r.startIndex) : console.warn(
        `Prop startIndex set to ${r.startIndex} is greater than the number of tabs - ${l.value.length}. Make sure that the starting index is less than the number of tabs registered`
      );
    };
    ee("addTab", ie), ee("removeTab", ce), te(() => r.startIndex, (e) => {
      e < l.value.length && j(e);
    });
    const ye = p("");
    return te(ye, (e) => {
      be(e);
    }), le(() => {
      he();
    }), (e, n) => (d(), u("div", {
      id: s(m),
      class: T(["vue-form-wizard", [t.stepSize, { vertical: s(y) }]]),
      onKeyup: [
        K(re, ["right"]),
        K(ue, ["left"])
      ]
    }, [
      e.$slots.title ? (d(), u("div", Ee, [
        v(e.$slots, "title", {}, () => [
          g("h4", Ke, z(t.title), 1),
          g("p", Oe, z(t.subtitle), 1)
        ])
      ])) : O("", !0),
      g("div", Pe, [
        s(y) ? O("", !0) : (d(), u("div", Ae, [
          g("div", {
            class: "wizard-progress-bar",
            style: k(s(J))
          }, null, 4)
        ])),
        g("ul", {
          class: T(["wizard-nav wizard-nav-pills", t.stepsClasses]),
          role: "tablist"
        }, [
          (d(!0), u(ge, null, ke(l.value, (c, f) => v(e.$slots, "step", {
            tab: c,
            index: f,
            navigateToTab: R,
            stepSize: t.stepSize,
            transition: t.transition
          }, () => [
            W(se, {
              tab: c,
              "step-size": t.stepSize,
              onClick: (E) => t.disableBackOnClickStep || t.disableBack ? !1 : R(f),
              onKeyup: K((E) => R(f), ["enter"]),
              transition: t.transition,
              index: f,
              "disable-back-on-click-step": t.disableBack ? !0 : t.disableBackOnClickStep
            }, null, 8, ["tab", "step-size", "onClick", "onKeyup", "transition", "index", "disable-back-on-click-step"])
          ])), 256))
        ], 2),
        g("div", Le, [
          v(e.$slots, "default", w($(s(B))))
        ])
      ]),
      t.hideButtons ? O("", !0) : (d(), u("div", Me, [
        v(e.$slots, "footer", w($(s(B))), () => [
          t.disableBack ? O("", !0) : (d(), u("div", Re, [
            s(L) ? (d(), u("span", {
              key: 0,
              onClick: V,
              onKeyup: K(V, ["enter"]),
              role: "button",
              tabindex: "0"
            }, [
              v(e.$slots, "prev", w($(s(B))), () => [
                W(q, {
                  style: k(s(M)),
                  disabled: C.value
                }, {
                  default: U(() => [
                    G(z(t.backButtonText), 1)
                  ]),
                  _: 1
                }, 8, ["style", "disabled"])
              ])
            ], 40, Fe)) : O("", !0),
            v(e.$slots, "custom-buttons-left", w($(s(B))))
          ])),
          g("div", Ne, [
            v(e.$slots, "custom-buttons-right", w($(s(B)))),
            s(S) ? (d(), u("span", {
              key: 0,
              onClick: A,
              onKeyup: K(A, ["enter"]),
              role: "button",
              tabindex: "0"
            }, [
              v(e.$slots, "finish", w($(s(B))), () => [
                W(q, {
                  style: k(s(M))
                }, {
                  default: U(() => [
                    G(z(t.finishButtonText), 1)
                  ]),
                  _: 1
                }, 8, ["style"])
              ])
            ], 40, We)) : (d(), u("span", {
              key: 1,
              onClick: A,
              onKeyup: K(A, ["enter"]),
              role: "button",
              tabindex: "0"
            }, [
              v(e.$slots, "next", w($(s(B))), () => [
                W(q, {
                  style: k(s(M)),
                  disabled: C.value
                }, {
                  default: U(() => [
                    G(z(t.nextButtonText), 1)
                  ]),
                  _: 1
                }, 8, ["style", "disabled"])
              ])
            ], 40, qe))
          ])
        ])
      ]))
    ], 42, xe));
  }
}), Ve = ["id", "aria-hidden", "aria-labelledby"], De = /* @__PURE__ */ H({
  props: {
    title: { default: "" },
    icon: { default: "" },
    customIcon: { default: "" },
    lazy: { type: Boolean, default: !1 },
    beforeChange: null,
    afterChange: null,
    route: null,
    additionalInfo: { default: () => ({}) }
  },
  setup(t) {
    const o = t, r = ae("addTab"), m = ae("removeTab"), a = p(!1), b = p(null), C = p(!1), l = p(""), h = me(), S = i(() => h?.parent?.props?.shape || "circle"), y = i(() => h?.parent?.props?.color || "#e74c3c"), L = i(() => h?.parent?.props?.errorColor || "#8b0000"), P = i(() => ({
      title: o.title,
      icon: o.icon,
      customIcon: o.customIcon,
      beforeChange: o.beforeChange,
      afterChange: o.afterChange,
      route: o.route,
      active: a.value,
      checked: C.value,
      validationError: b.value,
      tabId: l.value,
      color: y.value,
      errorColor: L.value,
      shape: S.value
    }));
    return le(() => {
      r && r(P.value);
    }), Ce(() => {
      m && m(P.value);
    }), (I, J) => !t.lazy || a.value ? Te((d(), u("div", {
      key: 0,
      class: "wizard-tab-container",
      role: "tabpanel",
      id: l.value,
      "aria-hidden": !a.value,
      "aria-labelledby": `step-${l.value}`
    }, [
      v(I.$slots, "default", { active: a.value }),
      v(I.$slots, "customIcon")
    ], 8, Ve)), [
      [pe, a.value]
    ]) : O("", !0);
  }
}), Ue = {
  install(t) {
    t.component("form-wizard", He), t.component("tab-content", De), t.component("wizard-button", q), t.component("wizard-step", se);
  }
};
export {
  He as FormWizard,
  De as TabContent,
  q as WizardButton,
  se as WizardStep,
  Ue as default
};
